src/mon/MonMap.h

```diff
@@ -25,34 +25,124 @@ namespace ceph {
   class Formatter;
 }
 
+struct mon_info_t {
+  /**
+   * monitor name
+   *
+   * i.e., 'foo' in 'mon.foo'
+   */
+  string name;
+  /**
+   * monitor's public address
+   *
+   * public facing address, traditionally used to communicate with all clients
+   * and other monitors.
+   */
+  entity_addr_t public_addr;
+
+  mon_info_t(string &n, entity_addr_t& p_addr)
+    : name(n), public_addr(p_addr)
+  { }
+
+  mon_info_t() { }
+
+
+  void encode(bufferlist& bl) const;
+  void decode(bufferlist::iterator& p);
+  void print(ostream& out) const;
+};
+WRITE_CLASS_ENCODER(mon_info_t)
+inline ostream& operator<<(ostream& out, const mon_info_t& mon) {
+  mon.print(out);
+  return out;
+}
+
+
 class MonMap {
  public:
   epoch_t epoch;       // what epoch/version of the monmap
   uuid_d fsid;
-  map<string, entity_addr_t> mon_addr;
   utime_t last_changed;
   utime_t created;
 
-  map<entity_addr_t,string> addr_name;
+  /* we don't use this anymore, but we keep it around to make
+   * compatibility easier, as there may be monitors that do not
+   * support our way to do monmaps (e.g., during an upgrade).
+   *
+   * This map is kept populated at all times with the public address of
+   * each entry in 'mons'.
+   *
+   * This will be encoded with every map we produce just as well. We could
+   * generate this during encoding, but seems like that's just wasting
+   * cycles. And this barelly wastes any resources, so there's no big
+   * tradeoff anyway.
+   */
+  map<string, entity_addr_t> mon_addr;
+
+  map<string, mon_info_t> mons;
+  map<entity_addr_t, mon_info_t> addr_mons;
+
+  vector<mon_info_t> ranks;
```

Comment:
i don't like having duplicate mon_info_t.  I would prefer a single map mon_info with the mon_info_t, and everythign else an index into that.  The problem I see is that we current determine rank by addr sort order, but we'll soon have multiple addrs. We can redefine that as public addr sort order, and make it map<entity_addr_t,mon_info_t> mon_info; I guess, but that's still a bit annoying.  Once all the mons understand the new encoding we could redefine rank order in a new, less inflexible way.

So maybe your vector<mon_info_t> mon_info is the right path.  But then everything else should be map<entity_addr_t,int>, map<string,int>, etc.

I don't really like duplicating the mon_addr *just* so we can encode it; I'd rather do the compat hijinx in encode/decode only.